File: aoc-kotlin/src/main/java/me/sol/Grid.java
package me.sol;

import aoc.Direction;

import java.io.*;
import java.util.Arrays;
import java.util.stream.IntStream;

public record Grid(long[][] rows) {
    public static final int OUT = -1;

    public int toCoordinate(int x, int y) {
        return y * width() + x;
    }

    public int getX(int coordinate) {
        return coordinate % width();
    }

    public int getY(int coordinate) {
        return coordinate / width();
    }

    public boolean inbounds(int x, int y) {
        return 0 <= x && x < rows[0].length && 0 <= y && y < rows.length;
    }

    public int move(int coordinate, Direction direction) {
        return move(coordinate, direction, 1);
    }

    public int move(int coordinate, Direction direction, int delta) {
        if (coordinate == OUT) {
            return OUT;
        }
        int x = getX(coordinate);
        int y = getY(coordinate);
        int newX = switch (direction) {
            case UP, DOWN -> x;
            case RIGHT -> x + delta;
            case LEFT -> x - delta;
        };
        int newY = switch (direction) {
            case UP -> y - delta;
            case RIGHT, LEFT -> y;
            case DOWN -> y + delta;
        };
        return inbounds(newX, newY) ? toCoordinate(newX, newY) : OUT;
    }

    public long getValue(int coordinate) {
        return rows[getY(coordinate)][getX(coordinate)];
    }

    public void setValue(int coordinate, long value) {
        rows[getY(coordinate)][getX(coordinate)] = value;
    }

    public int move(int origin, int deltaX, int deltaY) {
        int newX = getX(origin) + deltaX;
        int newY = getY(origin) + deltaY;
        return inbounds(newX, newY) ? toCoordinate(newX, newY) : OUT;
    }

    public IntStream neighbours(int origin) {
        return Arrays.stream(Direction.values())
                .mapToInt(d -> move(origin, d))
                .filter(n -> n != OUT);
    }

    public int getSurroundingCount(int origin, long value) {
        var count = 0;
        for (var deltaY = -1; deltaY <= 1; deltaY++) {
            for (var deltaX = -1; deltaX <= 1; deltaX++) {
                if (!(deltaX == 0 && deltaY == 0)) {
                    var coord = move(origin, deltaX, deltaY);
                    if (coord != OUT && getValue(coord) == value) {
                        count++;
                    }
                }
            }
        }
        return count;
    }

    public String coordToString(int coordinate) {
        return "(%d,%d)".formatted(getX(coordinate), getY(coordinate));
    }

    public int width() {
        return rows[0].length;
    }

    public int height() {
        return rows.length;
    }

    public int distance(int pos1, int pos2) {
        return Math.abs(getX(pos2) - getX(pos1)) + Math.abs(getY(pos2) - getY(pos1));
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == this) {
            return true;
        }
        if (obj == null || obj.getClass() != getClass()) {
            return false;
        }
        var that = (Grid) obj;
        return Arrays.deepEquals(rows, that.rows);
    }

    @Override
    public int hashCode() {
        return Arrays.hashCode(rows);
    }

    @Override
    public String toString() {
        return "Grid[lines=" + rows + ']';
    }

    public int[] findAll(long value) {
        return IntStream.range(0, height())
                .flatMap(y -> IntStream.range(0, width()).map(x -> toCoordinate(x, y)))
                .filter(c -> getValue(c) == value)
                .toArray();
    }

    IntStream streamCoordinates() {
        return IntStream.range(0, rows.length)
                .flatMap(y -> IntStream.range(0, rows[0].length).map(x -> toCoordinate(x, y)));
    }

    public void printAsChars(BufferedWriter writer) {
        try {
            for (int y = 0; y < rows.length; y++) {
                for (int x = 0; x < rows[0].length; x++) {
                    writer.write((int) rows[y][x]);
                }
                writer.newLine();
            }
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    public void printAsChars(OutputStream outputStream) {
        try (var writer = new BufferedWriter(new OutputStreamWriter(outputStream))) {
            printAsChars(writer);
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    public Grid copy() {
        return new Grid(Utility.copy(rows));
    }

    @Override
    public long[][] rows() {
        return Utility.copy(rows);
    }
}

File: aoc-kotlin/src/main/kotlin/aoc/Grid.kt
package aoc

data class Coordinate(val x: Int, val y: Int) {
    fun moveMirrored(direction: Direction): Coordinate {
        return when (direction) {
            Direction.UP -> Coordinate(x, y + 1)
            Direction.RIGHT -> Coordinate(x + 1, y)
            Direction.DOWN -> Coordinate(x, y - 1)
            Direction.LEFT -> Coordinate(x - 1, y)
        }
    }

    fun move(direction: Direction): Coordinate {
        return when (direction) {
            Direction.UP -> Coordinate(x, y - 1)
            Direction.RIGHT -> Coordinate(x + 1, y)
            Direction.DOWN -> Coordinate(x, y + 1)
            Direction.LEFT -> Coordinate(x - 1, y)
        }
    }
}

data class LongCoordinate(val x: Long, val y: Long) {
    fun move(direction: Direction): LongCoordinate {
        return when (direction) {
            Direction.UP -> LongCoordinate(x, y + 1L)
            Direction.RIGHT -> LongCoordinate(x + 1L, y)
            Direction.DOWN -> LongCoordinate(x, y - 1L)
            Direction.LEFT -> LongCoordinate(x - 1L, y)
        }
    }
}

enum class Direction {
    UP, RIGHT, DOWN, LEFT;

    fun opposite(): Direction {
        return Direction.entries.get((this.ordinal + 2) % 4)
    }
}

data class Coordinate3D(val x: Int, val y: Int, val z: Int) {
    fun move(direction: Direction3D): Coordinate3D {
        return when (direction) {
            Direction3D.UP -> Coordinate3D(x, y + 1, z)
            Direction3D.RIGHT -> Coordinate3D(x + 1, y, z)
            Direction3D.DOWN -> Coordinate3D(x, y - 1, z)
            Direction3D.LEFT -> Coordinate3D(x - 1, y, z)
            Direction3D.IN -> Coordinate3D(x, y, z - 1)
            Direction3D.OUT -> Coordinate3D(x, y, z + 1)
        }
    }

}

enum class Direction3D {
    UP, RIGHT, DOWN, LEFT, IN, OUT
}
